rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,369 @@
+# RFC 12217 - 2022-08-25 - Handling Discarded Events
+
+We want to give Vector users the assurance that events are never lost unexpectedly. There are
+currently many circumstances in which events are simply discarded without recourse for users of
+Vector to recover those events. This RFC presents a framework for handling those discarded events in
+a uniform fashion in all components.
+
+## Context
+
+- [Support dead letter queue sources](https://github.com/vectordotdev/vector/issues/1772)
+- [Vector sources acknowledge discarded events](https://github.com/vectordotdev/vector/issues/12217)
+- [Reroute discarded events from throttle transform](https://github.com/vectordotdev/vector/issues/13549)
+- Sink Error Handling RFC (internal document)
+
+## Scope
+
+### In scope
+
+- Handling of discarded and errored events in sources, transforms, and sinks.
+
+### Out of scope
+
+- Handling of discarded data in sources before it is translated into events.
+- Handling of discarded sink batch buffers.
+
+## Pain
+
+Vector has a number of ways in which components may drop events due that are outside of the
+operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
```

Comment:
Is this accurate? According to [the docs for `trait Message`](https://docs.rs/prost/latest/prost/trait.Message.html) that signature is `fn encode<B: BufMut>(&self, buf: &mut B)`.

Reply:
Hmmm.... I'm not sure why I was originally thinking it was a limitation on the `prost` side. I guess `Encodable::encode` takes `self` because we need to consume it for `proto::EventArray::from`.

The comment about not being currently applicable still applies, but at least easier to solve, potentially, than changing a trait upstream.

Good eye.